### PR TITLE
ENH: add two independent proportion inference rebased

### PR DESCRIPTION
**Pull request by josef-pkt**
rebased version of #4829

### CODE REVIEW

**Comment by josef-pkt** [comment]
rebased, one minor merge conflict, no additional changes compared to old PR

File: statsmodels/stats/proportion.py [hunk 1]
```diff
@@ -6,13 +6,20 @@
...
 import numpy as np
 from scipy import stats, optimize
 from sys import float_info
 
 from statsmodels.stats.base import AllPairsResults
 from statsmodels.tools.sm_exceptions import HypothesisTestWarning
+from statsmodels.stats.weightstats import _zstat_generic2
+
+
+class Holder(object):
```
Comment (josef-pkt) [inline statsmodels/stats/proportion.py:20]: use generic holder class now, currently tools/testing

File: statsmodels/stats/proportion.py [hunk 2]
```diff
@@ -1044,3 +1061,802 @@ def proportions_chisquare_pairscontrol(count, nobs, value=None,
                                    )[1]
                for pair in all_pairs]
     return AllPairsResults(pvals, all_pairs, multitest_method=multitest_method)
+
+
+def confint_proportion_2indep(count1, nobs1, count2, nobs2, method=None,
+                              compare='diff', alpha=0.05, correction=True):
```
Comment (josef-pkt) [inline statsmodels/stats/proportion.py:1067]: function looks fine to me when skimming the code

**Comment by josef-pkt** [comment]
code duplication

- conf_int, 
- hypothesis test, 
- effectsizes and variance for those in meta-analysis PR #6632
- power?

plus contingency table version for some of it.

This contains power function for normal with alternative dependent variance as extra enhancement.

**Comment by josef-pkt** [comment]
notebook for old PR is not up-to-date with code

return of `test_proportions_2indep` is now Holder instance and
`TypeError: 'Holder' object is not subscriptable`

**Comment by josef-pkt** [comment]
from notebook, this seems to be what's currently implemented in the test and confint functions

```
                           statistic    pvalue     lower      upper
compare    method                                                  
diff       agresti-caffo    2.106741  0.035140  0.011612   0.321722
           newcomb               NaN       NaN  0.018921   0.340369
           score            2.241651  0.024984       NaN        NaN
           wald             2.348087  0.018870  0.029169   0.323772
odds-ratio logit            1.951278  0.051024  0.990490  73.900302
           logit-adjusted   1.941318  0.052220  0.982796  37.748599
           logit-smoothed   1.950483  0.051119  0.990102  60.488353
           score            2.241651  0.024984       NaN        NaN
ratio      log              1.868966  0.061628  0.909606  53.869506
           log-adjusted     1.868374  0.061710  0.924135  27.052314
           score            2.241651  0.024984  1.208568  43.032972
```

**Comment by josef-pkt** [comment]
rebased after merging meta-analysis into master

File: statsmodels/stats/proportion.py [hunk 3]
```diff
@@ -1044,3 +1061,802 @@ def proportions_chisquare_pairscontrol(count, nobs, value=None,
...
+
+    Returns
+    -------
+    low, upp
+
+    Notes
+    -----
+    Status: experimental, API and defaults might still change.
+        Insufficiently tested, more ``methods`` will be added.
+
```
Comment (josef-pkt) [inline statsmodels/stats/proportion.py:1114]: References is missing
need description of methods

**Comment by josef-pkt** [comment]
codacy complains about some unused variables, 2 in proportions.py and several in unit tests

File: statsmodels/stats/proportion.py [hunk 4]
```diff
@@ -1044,3 +1061,802 @@ def proportions_chisquare_pairscontrol(count, nobs, value=None,
...
+            d_low = np.sqrt((p1 - low1)**2 + (upp2 - p2)**2)
+            d_upp = np.sqrt((p2 - low2)**2 + (upp1 - p1)**2)
+            low = diff - d_low
+            upp = diff + d_upp
+
+        else:
+            raise ValueError('method not recognized')
+
+    elif compare == 'ratio':
+        ratio = p1 / p2
```
Comment (josef-pkt) [inline statsmodels/stats/proportion.py:1161]: lint: not used

File: statsmodels/stats/proportion.py [hunk 5]
```diff
@@ -1044,3 +1061,802 @@ def proportions_chisquare_pairscontrol(count, nobs, value=None,
...
+        elif method == 'score':
+            res = _confint_riskratio_koopman(count2, count1, nobs2, nobs1,
+                                             alpha=alpha, correction=correction)
+            low, upp = res.confint
+
+        else:
+            raise ValueError('method not recognized')
+
+    elif compare == 'odds-ratio':
+        odds_ratio = p1 / (1 - p1) / p2 * (1 - p2)
```
Comment (josef-pkt) [inline statsmodels/stats/proportion.py:1184]: lint: not used

File: statsmodels/stats/tests/test_proportion.py [hunk 6]
```diff
@@ -574,3 +579,223 @@ def test_proportion_ztests():
...
+    odds-ratio
+    Woolf logit 0.99 74 4.31
+    Gart adjusted logit 0.98 38 3.65
+    Independence-smoothed logit 0.99 60 4.11
+    Cornfield exact conditional 0.97 397 6.01
+    Cornfield mid-p 1.19 200 5.12
+    Baptista–Pike exact conditional 1.00 195 5.28
+    Baptista–Pike mid-p 1.33 99 4.31
+    Agresti–Min exact unconditional 1.19 72 4.10
+    '''
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_proportion.py:617]: pointless-string-statement
try ` # pylint: disable=W0105`

File: statsmodels/stats/tests/test_proportion.py [hunk 7]
```diff
@@ -574,3 +579,223 @@ def test_proportion_ztests():
...
+def test_score_confint_koopman_nam():
+
+    #example Koopman, based on Nam 1995
+
+    x0, n0 = 16, 80
+    x1, n1 = 36, 40
+    x = x0 + x1
+    n = n0 + n1
+    p0 = x0 / n0
+    p1 = x1 / n1
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_proportion.py:743]: 4 unused variable

File: statsmodels/stats/proportion.py [hunk 8]
```diff
@@ -541,6 +555,7 @@ def binom_tost_reject_interval(low, upp, nobs, alpha=0.05):
     x_upp = stats.binom.ppf(alpha, nobs, upp) - 1
     return x_low, x_upp
 
+
```
Comment (josef-pkt) [inline statsmodels/stats/proportion.py:558]: those lines are pep-8 cleaning
not clean in master

**Comment by josef-pkt** [comment]
This PR is missing the equivalence TOST tests
this should work generically. i.e. just call `test_proportions_2indep` for the one-sided tests.
see NCSS documentation chapter 516

File: statsmodels/stats/proportion.py [hunk 9]
```diff
@@ -1044,3 +1061,802 @@ def proportions_chisquare_pairscontrol(count, nobs, value=None,
...
+            res.ratio = ratio
+            res.odds_ratio = odds_ratio
+        return res
+    else:
+        return statistic, pvalue
+
+
+def _std_2prop_power(diff, p2, ratio=1, alpha=0.05, value=0):
+    """compute standard error under null and alternative for 2 proportions
+
```
Comment (josef-pkt) [inline statsmodels/stats/proportion.py:1651]: only for "diff" ?
nobs_ratio would be more explicit, if it is ratio of nobs

Comment (josef-pkt) [reply statsmodels/stats/proportion.py:1651]: in power computation, I always use ratio for nobs-ratio.

but for 2sample proportion, we also have risk-ratio and odds-ratio

**Comment by josef-pkt** [comment]
from notebook: check consistency between confint and hypothesis test:
compute pvalue at upper confint bound

score method is inconsistent

```

                           statistic    pvalue
compare    method                             
diff       agresti-caffo   -1.959964  0.050000
           wald            -1.959964  0.050000
odds-ratio logit           -1.959964  0.050000
           logit-adjusted  -1.959964  0.050000
           logit-smoothed  -1.959964  0.050000
ratio      log             -1.959964  0.050000
           log-adjusted    -1.959964  0.050000
           score           -1.974536  0.048321
```

File: statsmodels/stats/proportion.py [hunk 10]
```diff
@@ -1044,3 +1061,802 @@ def proportions_chisquare_pairscontrol(count, nobs, value=None,
...
+            p2_ = count2_ / nobs2_
+            ratio_ = p1_ / p2_
+            var = (1 / count1_) - 1 / nobs1_ + 1 / count2_ - 1 / nobs2_
+            z = stats.norm.isf(alpha / 2)
+            d_log = z * np.sqrt(var)
+            low = np.exp(np.log(ratio_) - d_log)
+            upp = np.exp(np.log(ratio_) + d_log)
+
+        elif method == 'score':
+            res = _confint_riskratio_koopman(count2, count1, nobs2, nobs1,
```
Comment (josef-pkt) [inline statsmodels/stats/proportion.py:1176]: why is the argument sequence reversed `count2, count1`?

**Comment by josef-pkt** [comment]
unit test coverage

test_2proportions diff score and or score have no coverage
ratio score is covered

**Comment by josef-pkt** [comment]
what is  `_confint_riskratio_paired_nam(table, alpha=0.05)` ?
It says paired, but I thought I only have code for two independent samples

**Comment by josef-pkt** [comment]
current methods
I didn't check what's in power

| statistic |  method | confint | test | power |
|:--------- |:------- |:-------:|:----:|:------|
| diff | 'wald'           | Y | Y | ? |
|      | 'agresti-caffo'  | Y | Y | - |
|      | 'newcomb'        | Y | NA | - |
|        | score          | - | Y | - |
| | | | |
| ratio  | 'log'          | Y | Y | - |
|        | 'log-adjusted' | Y | Y | - |
|        | score          | Y | Y | - |
| | | | |
| odds-ratio | 'logistic' | Y | Y | - |
|            | 'logistic-adjusted' | Y | Y | - |
|            | 'logistic-smoothed' | - | Y | - |
|            | score | - | Y | - |


in **meta-analysis effect size**, 
we have wald diff, log risk ratio, log odds ratio and arcsine
no backtransform for log-rr and log-or
more flexible options for adding fake observations, i.e. continuity corrections.
includes standard errors

**Comment by josef-pkt** [comment]
This PR does not include any exact methods, see #6676 but only for equality testing (AFAIR).

All methods here are asymptotic with small sample adjustments, transformation and artificial observations.

**Comment by josef-pkt** [comment]
`score_test_proportion_2indep` is a public function.
is also included as option in `test_proportions_2indep`, just delegating to the score_test function

Returns:
`confint_proportion_2indep` 
It just returns confint (low, upp), most intermediate results are method specific, e.g. standard errors for log or logit transformed statistic, with or without continuity correction. newcomb method doesn't have any standard errors
I guess there is not much point in adding additional results

`test_proportions_2indep` and `score_test_proportion_2indep` return a Holder results instance by default, keyword `return_results=True`

`score_test_proportion_2indep` returns the constrained MLE for p1 and p2 in the holder instance

**Comment by josef-pkt** [comment]
score confidence interval
updated link https://github.com/josef-pkt/statsmodels/pull/5#issuecomment-135839951
https://ncss-wpengine.netdna-ssl.com/wp-content/themes/ncss/pdf/Procedures/PASS/Confidence_Intervals_for_the_Ratio_of_Two_Proportions.pdf

looks like rootfinding and not a direct implementation:
"Miettinen and Nurminen (1985) proposed inverting their score test to find the confidence interval. The lower limit is found by solving ..."

**Comment by josef-pkt** [comment]
valid input ranges

I used the wrong `value`, and variance is negative
v=59 is from risk ratio, but that doesn't work for "diff".

There should be a check that the user provides a valid `value`, or we truncate props to [0, 1]
Constrained MLE for score_test has negative proportion. 
However, we don't know in advance what are the feasible "diff" `value`s for constrained MLE.

```
# check 
rt0 = test_proportions_2indep(count1, nobs1, count2, nobs2, 
                                 value=v, compare="diff", method='score',
                                 alternative="two-sided")
vars(rt0)
{'statistic': nan,
 'pvalue': nan,
 'compare': 'diff',
 'method': 'score',
 'variance': -100.59951372870873,
 'alternative': 'two-sided',
 'prop1_null': 58.54406793461139,
 'prop2_null': -0.7123891759497276,
 'diff': 0.1764705882352941,
 'ratio': 7.0,
 'odds_ratio': 8.555555555555555}
v
59.256457110561115
```

**Comment by josef-pkt** [comment]
another issue: 
The currently implemented confint are always two-sided. There is no option for one-sided confint.
In this case, confint are symmetric in the underlying (transformed) random variable and we use normal distribution. So, one-sided is just one confint bound at alpha * 2

**Comment by josef-pkt** [comment]
another consistency check:
tost evaluated using equivalence margin from confint with alpha=0.05


```
                           pvalue
compare    method                
diff       agresti-caffo    0.025
           score            0.025
           wald             0.025
odds-ratio logit            0.025
           logit-adjusted   0.025
           logit-smoothed   0.025
           score            0.025
ratio      log              0.025
           log-adjusted     0.025
           score            0.025
```

**Comment by josef-pkt** [comment]
next question: Can I get the power generically, i.e. for any method available in the test and tost function?

the `power_proportion_2indep` function in this PR uses new `normal_power_het` for explicit computation but has only "diff" "wald" (?)

If I have the props under the alternative, then I could do the full binomial power version, but would need to evaluate (run hypothesis test on) all points on grid. That's what I was doing for "exact" power.
(simplest version would be when I have the rejection region in terms of counts, but that might not be easy to get analytically.)

If I cannot do it generically this way, then I need mean and variance of the test statistic under the alternative (i.e. noncentrality) for each method similar to the current one for diff-wald.

test_proportions_2indep works vectorized for all compare and methods
(why does constrained MLE work for score test in vectorized way?)
Not all confint will work vectorized because of the root finding (in score confint)

PASS Superiority by a Margin for two proportions chapter 205 (printed several years ago) uses
binomial exact power after computing rejection region in terms of counts.
Large sample power is based on normal distribution. However, "the results for the odds-ration have not (to our knowledge) been published. In this case, we substitute the calculations based on the ratio."

I guess my naming with `compare="ratio"` came from NCSS/PASS. Newer version as in meta-analysis uses "risk ratio" or "rr", and "rd" as abbreviation for risk difference ("diff").

**Comment by josef-pkt** [comment]
getting close, I have most of what I want in this round, I'm skipping power this time

now also equivalence TOST tests

'logit-smoothed' for odds-ratio is available in hypothesis test and tost but not in confint.

needs another check for method names
e.g. "agresti-caffo" should maybe become "wald-adjusted"
maybe we need "continuity correction", i.e. which "adjustment"
maybe wald-cc and wald-ac, or separate keyword that applies to wald tests

also need to check that we have consistent use of abbreviation for the compare statistic, e.g. "or" for "odds-ratio"

docs and references, e.g. do `methods` have explanation, author names?

corner cases, AFAIR, I never checked corner cases counts=0 or counts=nobs

test noise: I get some warnings when running the unit tests, e.g.
```
statsmodels/stats/tests/test_proportion.py::test_equivalence_2indep
  ...\python-3.7.4.amd64\lib\site-packages\scipy
\stats\_distn_infrastructure.py:1892: RuntimeWarning: invalid value encountered
in less_equal
    cond2 = cond0 & (x <= _a)
```

**Comment by josef-pkt** [comment]
silly naming mistake:
test_proportions_2indep, 

  proportion is singular:
confint_proportion_2indep, power_proportion_2indep score_test_proportion_2indep  samplesize_proportion_2indep_onetail 

name should be plural for 2 and k-sample tests

**Comment by josef-pkt** [comment]
replicating 2 sample tests using GLM works, but not with irls, see #3703

Note: endog in GLM binomial is (success, failure) and not (counts, nobs) as in stats.proportions.
Initially, I made a mistake and numbers didn't match up.


```
en = np.array([[count1, nobs1 - count1], [count2, nobs2 - count2]])
ex = np.array([[1., 1], [1, 0]])
mod_glm = GLM(en, ex, family=families.Binomial())
res_glm = mod_glm.fit(method='bfgs', tol=1e-10, max_start_irls=0)
res_glm.params[-1]
np.exp(res_glm.conf_int())[1]  # Wald confidence interval for odds-ratio
```

using identity link replicated `compare="diff"`

**Comment by josef-pkt** [comment]
azure py 3.6 failed in  `statsmodels/tsa/statespace/tests/test_exact_diffuse_filtering.py::test_nondiagonal_obs_cov`

this PR is green

**Comment by josef-pkt** [comment]
`eform` after glm, function has been merged already

```
r = test_proportions_2indep(count1, nobs1, count2, nobs2, compare="odds-ratio", method="logit")
vars(r)
{'statistic': 1.9512781246582152,
 'pvalue': 0.05102396984084286,
 'compare': 'odds-ratio',
 'method': 'logit',
 'diff': 0.1764705882352941,
 'ratio': 7.0,
 'odds_ratio': 8.555555555555555,
 'variance': 1.2101972101972103,
 'alternative': 'two-sided',
 'value': 1}

confint_proportions_2indep(count1, nobs1, count2, nobs2, compare="odds-ratio", method="logit")
(0.9904902816539896, 73.90030192115287)
```

```
np.exp(res_glm.conf_int())[1]
array([ 0.99049029, 73.90030122])
```

```
from statsmodels.genmod._prediction import params_transform_univariate
​
pres = params_transform_univariate(res_glm.params, res_glm.cov_params())
print(pres.summary_frame().loc[1:])
       mean   mean_se  mean_ci_lower  mean_ci_upper
1  8.555568  9.411897       0.990491      73.900488
```

**Comment by josef-pkt** [comment]
standard errors for two sample proportions ? and/or for transformed statistics

None of the functions in this PR provides standard errors. 
Using GLM we get standard errors for the parameters, e.g. log-odds-ratio, and backtransformed exp (and predict/link) versions.
In meta_analysis, we have standard errors for the effect size, which are transformed statistics like log-odds-ratio and log-risk ratio, but not for the backtransformed statistics.
In contingency_tables we have both, transformed and backtransformed statistics.

Can we get everything together in one place?

Not now, or at most partially.
The point here is to be able to add several methods to a single functions, e.g. confint.
Adding all functions together makes it more difficult to add idiosyncratic methods, e.g. `newcomb` mover confidence interval doesn't have analogues for other functions.

maybe a wrapper class or function to provide summary for params, hypo test and confint.

**Comment by josef-pkt** [comment]
great when things just work

GLM fit_constrained + score_test  == 2 proportion test with score and correction=False (Farrington-Manning)

```
v0 = np.log(0.8)
v0
-0.2231435513142097

mod_glm = GLM(en, ex, family=families.Binomial())
res_glm = mod_glm.fit_constrained("x1=-0.2231435", method='bfgs', tol=1e-10, max_start_irls=0)
print(
print(res_glm.summary())
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:           ['y1', 'y2']   No. Observations:                    2
Model:                            GLM   Df Residuals:                        1
Model Family:                Binomial   Df Model:                            0
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -6.3191
Date:                Sun, 03 May 2020   Deviance:                       7.0896
Time:                        15:37:09   Pearson chi2:                     6.58
No. Iterations:                     1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.9081      0.377     -5.066      0.000      -2.646      -1.170
x1            -0.2231          0       -inf      0.000      -0.223      -0.223
==============================================================================

Model has been estimated subject to linear equality constraints.


mod_glm.score_test(res_glm.params, k_constraints=1)
(array([6.57854016]), array([0.01032155]), 1)  # pvalue same as below at 7 decimals


r = test_proportions_2indep(count1, nobs1, count2, nobs2, value=0.8, compare="odds-ratio", method="score", correction=False)
vars(r)
{'statistic': 2.564867604246601,
 'pvalue': 0.010321518062695179,
 'compare': 'odds-ratio',
 'method': 'score',
 'variance': 0.5715431729016758,
 'alternative': 'two-sided',
 'prop1_null': 0.10609821391325566,
 'prop2_null': 0.1291959037338032,
 'diff': 0.1764705882352941,
 'ratio': 7.0,
 'odds_ratio': 8.555555555555555,
 'value': 0.8}

res_glm.predict()
array([0.10609832, 0.12919602])    # prop_null under constraint
```

**Comment by josef-pkt** [comment]
Holdertuple doesn't work by default for other than test results, e.g. power, confint. Also power returns a single value if not return_results.
Holder class will get repr and str in another open PR. (in robust/anova PR ?)

I'm about ready to stop and merge. 
One more check to see if I forgot about something.

File: statsmodels/stats/proportion.py (hunk 10 above)
Comment (josef-pkt) [reply statsmodels/stats/proportion.py:1178]: changed

**Comment by josef-pkt** [comment]
I will merge when green

**Comment by josef-pkt** [comment]
I still only have a partial overview of what we have for proportions # 6399

partial overlap with existing functions

| sample  | method | function(s) |
| --------  | --------  | ----------- |
| | value hyp | test_proportions_2indep <BR> proportions_ztest |
| | equivalence | tost_proportions_2indep <BR> proportions_ztost |

**Comment by josef-pkt** [comment]
completely green, merging
